datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,820 @@
+            aggregate_arguments,
+            group_by,
+            aggregator: ag,
+        })
+    }
+}
+
+pub fn new_group_values(
+    limit: usize,
+    sort: SortOptions,
+    vt: DataType,
+) -> Result<Box<dyn LimitedAggregator>> {
+    macro_rules! downcast_helper {
+        ($vt:ty, $d:ident) => {
+            return Ok(Box::new(PrimitiveAggregator::<$vt>::new(
+                limit,
+                limit * 10,
+                sort,
+            )))
+        };
+    }
+
+    downcast_primitive! {
+        vt => (downcast_helper, vt),
+        _ => {}
+    }
+
+    Err(DataFusionError::Execution(format!(
+        "Can't group type: {vt:?}"
+    )))
+}
+
+impl RecordBatchStream for GroupedTopKAggregateStream {
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+}
+
+pub trait LimitedAggregator: Send {
+    fn intern(&mut self, ids: ArrayRef, vals: ArrayRef) -> Result<()>;
+    fn emit(&mut self) -> Result<Vec<ArrayRef>>;
+    fn is_empty(&self) -> bool;
+}
+
+pub trait ValueType: ArrowNativeTypeOp + Clone {}
+
+impl<T> ValueType for T where T: ArrowNativeTypeOp + Clone {}
+
+pub trait KeyType: Clone + Eq + Hash {}
+
+impl<T> KeyType for T where T: Clone + Eq + Hash {}
+
+struct PrimitiveAggregator<VAL: ArrowPrimitiveType>
+where
+    <VAL as ArrowPrimitiveType>::Native: Clone,
+{
+    priority_map: PriorityMap<Option<String>, VAL::Native>,
+}
+
+impl<VAL: ArrowPrimitiveType> PrimitiveAggregator<VAL>
+where
+    <VAL as ArrowPrimitiveType>::Native: Clone,
+{
+    pub fn new(limit: usize, capacity: usize, sort: SortOptions) -> Self {
+        Self {
+            priority_map: PriorityMap::new(limit, capacity, sort),
+        }
+    }
+}
+
+unsafe impl<VAL: ArrowPrimitiveType> Send for PrimitiveAggregator<VAL> where
+    <VAL as ArrowPrimitiveType>::Native: Clone
+{
+}
+
+impl<VAL: ArrowPrimitiveType> LimitedAggregator for PrimitiveAggregator<VAL>
+where
+    <VAL as ArrowPrimitiveType>::Native: Clone,
+{
+    fn intern(&mut self, ids: ArrayRef, vals: ArrayRef) -> Result<()> {
+        let ids = ids.as_any().downcast_ref::<StringArray>().ok_or_else(|| {
+            DataFusionError::Execution("Expected StringArray".to_string())
+        })?;
+        let sort = self.priority_map.sort;
+        let vals = vals.as_primitive::<VAL>();
+        for row_idx in 0..ids.len() {
+            let val = if vals.is_null(row_idx) {
+                None
+            } else {
+                Some(vals.value(row_idx))
+            };
+            let id = if ids.is_null(row_idx) {
+                None
+            } else {
+                // Check goes here, because it is generalizable between str/String and Row/OwnedRow
+                let id = ids.value(row_idx);
+                if let Some(worst) = self.priority_map.worst_val() {
+                    if cmp(&sort, &val, worst) == Ordering::Greater {
+                        continue;
+                    }
+                }
+                Some(id.to_string())
```

Comment:
Any chance to remove the allocation and moving it to `insert`?

Reply:
Also if `ID` is not a string, it seems we currently needlessly convert it to use a string (and thus also allocation/hashing/etc. is slower)?

Reply:
Now that you have more context on the issue & the scope of the PR, does the present code:

```
                if self.priority_map.is_full() {
                    if self.priority_map.desc {
                        if let Some(worst) = self.priority_map.min_val() {
                            if val < *worst {
                                continue;
```

Satisfy your concerns?